File: src/main/java/com/mcommings/campaigner/modules/calendar/entities/CelestialEvent.java
package com.mcommings.campaigner.modules.calendar.entities;

import com.mcommings.campaigner.modules.common.entities.Campaign;
import jakarta.persistence.*;
import lombok.*;

@Getter
@Setter
@Builder
@Entity
@Table(name = "celestial_events",
        uniqueConstraints = {
                @UniqueConstraint(
                        columnNames = {"name", "fk_campaign_uuid"}
                )
        })
@NoArgsConstructor
@AllArgsConstructor
public class CelestialEvent {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(nullable = false, unique = true)
    private String name;

    private String description;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_campaign_uuid", nullable = false)
    private Campaign campaign;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_moon")
    private Moon moon;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_sun")
    private Sun sun;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_month", nullable = false)
    private Month month;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_week", nullable = false)
    private Week week;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fk_day", nullable = false)
    private Day day;

    @Column(name = "event_year", nullable = false)
    private int year;

}

File: src/main/frontend/lib/models/calendar/celestial_event.dart
class CelestialEvent {
  final int id;
  final String name;
  final String description;
  final String fkCampaignUuid;
  final int fkMoon;
  final int fkSun;
  final int fkMonth;
  final int fkWeek;
  final int fkDay;
  final int eventYear;

  const CelestialEvent({
    required this.id,
    required this.name,
    required this.description,
    required this.fkCampaignUuid,
    required this.fkMoon,
    required this.fkSun,
    required this.fkMonth,
    required this.fkWeek,
    required this.fkDay,
    required this.eventYear,
  });

  factory CelestialEvent.fromJson(Map<String, dynamic> json) {
    return switch (json) {
      {
        'id': int id,
        'name': String name,
        'description': String description,
        'fk_campaign_uuid': String fkCampaignUuid,
        'fk_moon': int fkMoon,
        'fk_sun': int fkSun,
        'fk_month': int fkMonth,
        'fk_week': int fkWeek,
        'fk_day': int fkDay,
        'event_year': int eventYear,
      } =>
        CelestialEvent(
          id: id,
          name: name,
          description: description,
          fkCampaignUuid: fkCampaignUuid,
          fkMoon: fkMoon,
          fkSun: fkSun,
          fkMonth: fkMonth,
          fkWeek: fkWeek,
          fkDay: fkDay,
          eventYear: eventYear,
        ),
      _ => throw const FormatException("Failed to load CelestialEvent."),
    };
  }
}
